Send logs to journal

ctx opened the pull request.

ctx committed:
Send logs to journal

gaudenz commented on tasks/journald-logging.yml:
```diff
@@ -0,0 +1,69 @@
+- name: Override dependencies for haproxy
+  template:
+    src: override.conf.j2
+    dest: /etc/systemd/system/haproxy.service.d/override.conf
```
Minor: To make the purpose of this override more explicit I would call it something like `journal-logging.conf`

gaudenz commented on tasks/journald-logging.yml:
```diff
@@ -0,0 +1,69 @@
+  loop:
+    - syslog.socket
+    - rsyslog
+    - haproxy
```
This should be done using handlers. And the services should also be enabled.

gaudenz commented on tasks/journald-logging.yml:
```diff
@@ -0,0 +1,69 @@
+  systemd:
+    name: var-lib-haproxy-dev-log.mount
+    state: stopped
+  when: dev_log_mount.stat.isreg is defined and dev_log_mount.stat.isreg
```
Why do we need this? I would prefer to roll-out the configuration and then (re)start (and enable) services.

I see that you have to replace the socket created by `rsyslog` by a file which is usually hidden by the bind mount. But I think the solution proposed below should work for this and does not have the negative consequence of stopping haproxy and logging daemons unconditionally while this role is executed.

gaudenz commented on templates/override.conf.j2:
```diff
@@ -0,0 +1,5 @@
+# Mount systemd logging socket
+
+[Unit]
+After=network.target rsyslog.service var-lib-haproxy-dev-log.mount
```
You don't have to list `network.target` and `rsyslog.service` as those are already part of the original file and this is additive.

gaudenz commented on tasks/journald-logging.yml:
```diff
@@ -0,0 +1,69 @@
+  file:
+    path: /var/lib/haproxy/dev/log
+    mode: 0644
+    state: touch
```
This is only needed the first time this role is executed. I would put this into a block only run under the condition that `/var/lib/haproxy/dev/log` is not mounted (has to be checked by a command unless you find an Ansible fact for it. I did not find one).

gaudenz reviewed (commented):
The most important thing is that this should use handlers to only restart services as needed. This is most important for haproxy as otherwise service will be interrupted even if there is no change at all.

ctx committed:
Restart services with handlers

gaudenz commented on tasks/journald-logging.yml:
```diff
@@ -0,0 +1,52 @@
+  template:
+    src: journal-logging.conf
+    dest: /etc/systemd/system/haproxy.service.d/journal-logging.conf
+  notify: restart and daemon_reload haproxy
```
As only the service dependencies are changed, there is no need to restart haproxy. A `daemon-reload` is enough.

gaudenz commented on tasks/journald-logging.yml:
```diff
@@ -0,0 +1,52 @@
+  template:
+    src: var-lib-haproxy-dev-log.mount
+    dest: /etc/systemd/system/var-lib-haproxy-dev-log.mount
+  notify: restart and daemon_reload haproxy
```
This should start the bind mount unit not restart haproxy.

gaudenz commented on handlers/main.yml:
```diff
@@ -6,3 +6,10 @@
     name: haproxy
     state: reloaded
+
+- name: restart and daemon_reload haproxy
+  systemd:
+    name: haproxy
+    enabled: yes
+    daemon_reload: true
+    state: restarted
```
I think you need two handlers. One to just reload the daemon and another to restart the mount unit (not haproxy).

gaudenz commented on tasks/journald-logging.yml:
```diff
@@ -0,0 +1,52 @@
+ name: Remove listen socket from haproxy config
```
A `-` is missing here. This is invalid YAML.

gaudenz commented on tasks/journald-logging.yml:
```diff
@@ -0,0 +1,52 @@
+    mode: 0644
+    state: touch
+  when: mount_status.rc is undefined or mount_status.rc != 0
+  notify: restart and daemon_reload haproxy
```
I don't think any restarting is needed here.

gaudenz reviewed (commented):
This now works and the result is like it should be but it still restarts haproxy when this is not needed.

ctx committed:
Reorganize handlers

gaudenz committed:
Restart bind mount and reload daemon on changes

This ensures that changes are also applied if the bound mount already
exists and that the changes are picked up by systemd even if the haproxy
config drop-in is not changed at the same time.

gaudenz commented:
I added one small change to not only start but restart the mount point on changes. LGTM now.
